tests/FormTest.php

```diff
@@ -43,35 +85,96 @@ public function testFormSubmit()
 
             // field has default, but form didn't send value back
-            $this->assertEquals(null, $f->model['name']);
+            $this->assertEquals(null, $m['name']);
 
-            $this->assertEquals('john@yahoo.com', $f->model['email']);
+            $this->assertEquals('john@yahoo.com', $m['email']);
 
             // security check, unspecified field must not be changed
-            $this->assertEquals(false, $f->model['is_admin']);
+            $this->assertEquals(false, $m['is_admin']);
+        });
+    }
 
-            $submit_called = true;
+    public function testTextArea()
+    {
+        $this->f->addField('TextArea');
+        $this->assertSubmit(['TextArea'=>'0'], function ($m) {
+            $this->assertSame('0', $m['TextArea']);
         });
+    }
 
-        $f->render();
-        unset($f);
-        $_POST = [];
+    public function assertSubmitError(array $post, callable $error_callback)
+    {
+        $this->assertSubmit($post, null, $error_callback);
+    }
+
+    public function assertFieldError(string $field, string $error)
+    {
+        $matched = false;
+
+        preg_replace_callback('/form\("add prompt","([^"]*)","([^"]*)"\)/', function ($matches) use ($error, $field, &$matched) {
+            if ($matches[1] == $field) {
+                $this->assertContains($error, $matches[2], 'Regarding field '.$field.' error message');
+
+                $matched = true;
+            }
+        }, $this->f_error);
+
+        $this->assertTrue($matched, 'Field '.$field.' did not produce error');
+    }
 
-        $this->assertTrue($submit_called);
+    public function assertFieldNoErrors(string $field)
+    {
+        preg_replace_callback('/form\("add prompt","([^"]*)","([^"]*)"\)/', function ($matches) use ($field, &$matched) {
+            if ($matches[1] == $field) {
+                $this->fail('Field '.$field.' unexpected error: '.$matches[2]);
+            }
+        }, $this->f_error);
+    }
+
+    public function testSubmitError()
+    {
+        $m = new Model();
+
+        $options = ['0'=>'yes please', '1'=>'woot'];
+
+        $m->addField('opt1', ['values'=>$options]);
+        $m->addField('opt2', ['values'=>$options]);
+        $m->addField('opt3', ['values'=>$options, 'required'=>true]);
+        //$m->addField('opt3_zerotest', ['values'=>$options, 'required'=>true]);
+        $m->addField('opt4', ['values'=>$options, 'mandatory'=>true]);
+
+        $this->f->setModel($m);
+        $this->assertSubmitError(['opt1'=>'2', 'opt3_zerotest'=>'0'], function ($error) {
+
+            // dropdown validates to make sure option is proper
+            $this->assertFieldError('opt1', 'not one of the allowed values');
+
+            // user didn't select any option here
+            $this->assertFieldNoErrors('opt2');
+
+            // dropdown insists for value to be there
+            $this->assertFieldError('opt3', 'Must not be null');
+
+            // value with '0' is valid selection
```

Comment:
@DarkSide666 - here is a test for dropdown value with key '0' which currently is considered invalid incorrect selection.